File: jwt_secret_test/create_jwt_client_secret.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// JWTClaims represents the payload structure for the client_secret JWT
type JWTClaims struct {
	ClientID string `json:"client_id"`
	jwt.RegisteredClaims
}

func createJWTClientSecret(clientID, clientSecret string, expiryMinutes int) (string, error) {
	// Calculate expiry time
	expiryTime := time.Now().Add(time.Duration(expiryMinutes) * time.Minute)

	// Create claims
	claims := JWTClaims{
		ClientID: clientID,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(expiryTime),
		},
	}

	// Create token with HS256 algorithm
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)

	// Sign the token with the client_secret
	tokenString, err := token.SignedString([]byte(clientSecret))
	if err != nil {
		return "", fmt.Errorf("failed to sign JWT: %w", err)
	}

	return tokenString, nil
}

func main() {
	// Check if correct number of arguments provided
	if len(os.Args) != 4 {
		fmt.Println("Usage: go run create_jwt_client_secret.go <client_id> <client_secret> <redirect_uri>")
		fmt.Println("Example: go run create_jwt_client_secret.go 26787adf-82fd-4838-b790-fbded3057755 '$2a$10$IvVS16zgyNYl77BF26.9zOQSaJeLSStxK20csim5H2OFXJhRofnAW' 'https://myapp.com/callback'")
		os.Exit(1)
	}

	// Read arguments
	clientID := os.Args[1]
	clientSecret := os.Args[2]
	redirectURI := os.Args[3]

	// Create JWT client_secret
	jwtClientSecret, err := createJWTClientSecret(clientID, clientSecret, 5)
	if err != nil {
		fmt.Printf("Error creating JWT: %v\n", err)
		os.Exit(1)
	}

	// Display results
	fmt.Println("=== JWT Client Secret Generator (Go) ===")
	fmt.Printf("Client ID: %s\n", clientID)
	fmt.Printf("Client Secret: %s\n", clientSecret)
	fmt.Printf("Redirect URI: %s\n", redirectURI)
	fmt.Printf("JWT Client Secret: %s\n\n", jwtClientSecret)

	// Decode and display JWT payload for verification
	token, err := jwt.ParseWithClaims(jwtClientSecret, &JWTClaims{}, func(token *jwt.Token) (interface{}, error) {
		return []byte(clientSecret), nil
	})

	if err != nil {
		fmt.Printf("Error parsing JWT: %v\n", err)
		os.Exit(1)
	}

	if claims, ok := token.Claims.(*JWTClaims); ok && token.Valid {
		fmt.Println("=== JWT Payload (Decoded) ===")
		payload := map[string]interface{}{
			"client_id": claims.ClientID,
			"exp":       claims.ExpiresAt.Unix(),
		}
		payloadJSON, _ := json.MarshalIndent(payload, "", "  ")
		fmt.Println(string(payloadJSON))
		fmt.Println()

		// Show expiry information
		fmt.Println("=== Expiry Information ===")
		fmt.Printf("Current Time: %s\n", time.Now().Format(time.RFC3339))
		fmt.Printf("Expiry Time: %s\n", claims.ExpiresAt.Format(time.RFC3339))
		// Use Unix timestamp for time calculation
		expTime := time.Unix(claims.ExpiresAt.Unix(), 0)
		fmt.Printf("Time Remaining: %s\n\n", time.Until(expTime))
	}

	// Print the OAuth2 integration URL
	fmt.Println("=== OAuth2 Integration URL ===")
	integrationURL := fmt.Sprintf("https://storx.io/oauth2-integration?client_id=%s&client_secret=%s&redirect_uri=%s&scope=read,write",
		clientID, jwtClientSecret, redirectURI)
	fmt.Println(integrationURL)
	fmt.Println()

}

File: jwt_secret_test/create_jwt_client_secret.py
#!/usr/bin/env python3
"""
JWT Client Secret Generator for StorX OAuth2 Token Request

This script creates a JWT client_secret according to the OAuth2 token workplan:
- Header: {"alg": "HS256", "typ": "JWT"}
- Payload: {"client_id": "...", "exp": unix_timestamp}
- Signed with the actual client_secret as HMAC key
"""

import json
import time
import jwt
from datetime import datetime, timedelta

def create_jwt_client_secret(client_id, client_secret, expiry_minutes=5):
    """
    Create a JWT client_secret for OAuth2 token request.
    
    Args:
        client_id (str): The client ID
        client_secret (str): The actual client secret used as HMAC key
        expiry_minutes (int): JWT expiry time in minutes (default: 5)
    
    Returns:
        str: JWT token string
    """
    
    # Calculate expiry timestamp (current time + expiry_minutes)
    expiry_timestamp = int(time.time()) + (expiry_minutes * 60)
    
    # JWT payload as specified in the workplan
    payload = {
        "client_id": client_id,
        "exp": expiry_timestamp
    }
    
    # Create JWT with HS256 algorithm
    jwt_token = jwt.encode(
        payload,
        client_secret,
        algorithm="HS256",
        headers={"typ": "JWT"}
    )
    
    return jwt_token

def main():
    # Client credentials from the provided data
    client_id = "e45fa79a-05f5-4f00-bbfe-bd0a14aead0a"
    client_secret = "2b4719e00623ad66c4cde8681efc59e9d40bf96cdb43d4b9cf859e27d5252102"
    
    # Create JWT client_secret
    jwt_client_secret = create_jwt_client_secret(client_id, client_secret)
    
    # Display results
    print("=== JWT Client Secret Generator ===")
    print(f"Client ID: {client_id}")
    print(f"Client Secret: {client_secret}")
    print(f"JWT Client Secret: {jwt_client_secret}")
    print()
    
    # Decode and display JWT payload for verification
    try:
        decoded_payload = jwt.decode(jwt_client_secret, client_secret, algorithms=["HS256"])
        print("=== JWT Payload (Decoded) ===")
        print(json.dumps(decoded_payload, indent=2))
        print()
        
        # Show expiry information
        exp_timestamp = decoded_payload["exp"]
        exp_datetime = datetime.fromtimestamp(exp_timestamp)
        current_time = datetime.now()
        
        print("=== Expiry Information ===")
        print(f"Current Time: {current_time}")
        print(f"Expiry Time: {exp_datetime}")
        print(f"Time Remaining: {exp_datetime - current_time}")
        print()
        
    except jwt.ExpiredSignatureError:
        print("ERROR: JWT has expired!")
    except jwt.InvalidTokenError as e:
        print(f"ERROR: Invalid JWT: {e}")
    
    # Example curl request
    print("=== Example curl Request ===")
    print("""curl -X POST \\
  http://localhost:10100/api/v0/oauth2/token \\
  -H "Content-Type: application/json" \\
  -d '{
    "client_id": "%s",
    "client_secret": "%s",
    "redirect_uri": "https://myapp.com/callback",
    "code": "AUTH_CODE_FROM_CONSENT",
    "passphrase": "your-passphrase"
  }'""" % (client_id, jwt_client_secret))

if __name__ == "__main__":
    main() 